### PR TITLE
Smallpatches

### PR DESCRIPTION
**Pull request by InfusOnWoW**

### CODE REVIEW

File: WeakAuras/GenericTrigger.lua [hunk 1]
```diff
@@ -2496,17 +2550,28 @@ function WeakAuras.WatchUnitChange(unit)
...
+        if newGuid ~= unitChangeGUIDS[unit] then
+          WeakAuras.ScanEvents("UNIT_CHANGED_" .. unit, unit)
           unitChangeGUIDS[unit] = newGuid
         end
+      else
+        for unit, guid in pairs(unitChangeGUIDS) do
+          local newGuid = WeakAuras.UnitExistsFixed(unit) and UnitGUID(unit) or ""
+          if guid ~= newGuid then
+            WeakAuras.ScanEvents("UNIT_CHANGED_" .. unit, unit)
+            print("ScanEvents ", "UNIT_CHANGED_", unit)
```
Comment (mrbuds) [inline WeakAuras/GenericTrigger.lua]: pritn left here

**Comment by mrbuds** [comment]
Status > Health trigger doesn't have the "unit" property in aura_env.state, so it can't be use with dynamic group "group by frame" options

**Comment by mrbuds** [comment]
with "health > party" in a dynamic group on raid frames, on group roster update i get some wrong clones with 1 health
https://i.imgur.com/3m8oOsy.jpg
```
!DF1wVTTnu4)lfyaRalds(ss6d7bBh5ed4iNAk30ouihkrAzUkt6rrzhNhYV9DoKs2YjPPbBadddiXGIINl889DUOy)4Oysrm5c(MyYQycd)d2IQthZLzMLXHTp1lMKO0mUUpn9BmTADmPFU4HhOA2J3fPu5gbS19twSOGBIdHJVRX6mTGfTBnpMm9IAfnqLR0GDc9XJe2Y(BB7VDWnnXKYcEpz6sL(gUEMuyI7dEvQrOKfwF2q1gybCYfcPOyPBTapjUcwB0ISmUUWzNO9BGlz8KYflCU1vbJVz4SXW7Tpsl10wGbwZZZhXkC6TOmHVHlneqkX9XK5d6rIMtI6nfuwjAuY6C6oq5ejDfVskRiGb40CisUxj3O5wLqUjy8y7DvEW1moNpvLxUsswtt5O3dxmdxlP5FcUqqmaIzNAJrJfRaJNaUDUitgtgeegfm1gFSkeraWUt5zvroq35kQfKbXNNMtlkq5H7BkU5QsaoRdHU3(0DlepWF6MW2jvSJMGRVfC9TGRVfC9(1UwDWZxCJsGHN(tIIMCnCdKIvud36mP0C392XxW9CRcyzagr(ZsQM)4DdlZZF8UBxkaXiA7D0HPSDamisZ0QY11Isq3oeHwfqDisLefsT9qqgo(wixyYTHy8WUlC8CBioSlcksiUkndbIiCfjt7DXOzeiqkG8gvU)2BE)TTdlhxBX6SGowrxiYCXFJkDtnmEUxBpaIqfeY36YwUvWWepWGl0azc4Tud16GuBgXqCx39KmyAqqisAnWlMS2CaLDUWiP1d8XBQHAS20RwrvaqnNHEu(gzw4OOHt7DDG17zIg6wtzIYci645vJBOMBKxsyqAeU5Cxwvbh1rrDsMn4dkAfviF7NEFM(B78ok8(0Far1FRk267v9SdI(8ZQA9IbiRetBghpANRvmW49MfnXLmGSpjRr6WWUN9HwNFA3oaO35d(ENhV4KUDBu)7OuKxR(fM6wSuTTxcqMtWAJ1rKyI)3RsZXj84dRvB5ABml5vRjcOQbW7FqzqqHuR)yde9Fw9pNPjtNC5SGdxHrYu1kHmdpfkuPSsmGtREADyS4kuO1ur)xQkmcOsHvzh)UK91Lbc5UJklJhU01s5jfFpsel9bB5SHh5oH9EfEcqEWwnaLBE6RwRZ31s6fQC3O8xn5kHQFELZN1wKXlaOqxxN8af(lpVuFfRzFxnWcZH)b1aQgFd)EdabuMA7NRP((7H8Ye89vNYT8Nwx94B4ABp3cfEdhQfp84DFSKYWczWidhB73WWano9xQ90t8R3)qS0L7T30ossOTqVDZnIcrsULBAF(OS8rHHbtNpoyyuD6)xAuqWE()Oe4Bl2Da02BkxZf)w1I(5gIAR8YJSHWWjHbia0(Vhai)FcaK89bGPJU8Q)fqay1sUiBjIfEhMj6F(yq2CY(5CjRQFWfx4MMLH(oNXzxtVhdah1CV7l2d))WTwFYvcSvYXJGvvX54pmqbd(aZjVV603N)Alrz7z6539SZAFMxRt745DElSNzNgICGLwxuUqGX9kEKfoSDdWzSgCv)1dAV7JxD5LvV6QgKaKlvQ5h(8MQZm4hNT9GsTYn4a21Q)RnK3J3161MKlF9s3WEBDJQ8H6XmUsWyCOaqyWNqIFZjkfs0whAjevFrgHnFwatY(1buiXrM1NQ)6SrNC4Pti2aerPfqx3QriUAY0r)(KWOEJFYSFUwkPl5PFR5ejO7UHQfum3MixNochig6w8B)Mn)zjvMX3lVPov5f(Eoctnpfivq0ecIvRiFsOPzW3OIFS58Em2fqoZpJeS5C5MF5X7Ex9639(x7lcr2beJSR3a(sNwi2f)x
```

**Comment by mrbuds** [comment]
Cast Trigger:
some properties are missing from aura_env.state: sourceName, sourceUnit (alias of unit), destName

**Comment by mrbuds** [comment]
Cast trigger: caster's target condition set to = player doesn't update on target changed

**Comment by mrbuds** [comment]
Cast trigger > unit = player, remaining time < 1

cast bar doesn't hide if i interrupt before end of cast (work fine if i don't check remaining time)

**Comment by mrbuds** [comment]
Kinda unrelated but while checking the "Assigned Role" property it made me remember it's not disable for classic in the Bufftrigger2 trigger

**Comment by mrbuds** [comment]
On classic:
```
3x WeakAuras\GenericTrigger.lua:2630: attempt to call global 'UnitGroupRolesAssigned' (a nil value)
WeakAuras\GenericTrigger.lua:2630: in function <WeakAuras\GenericTrigger.lua:2597>
[C]: ?
```

**Comment by mrbuds** [comment]
cast trigger: bars timers are inverted compared to previous behavior, casting a heal was going from 0 to 100% is not showing a bar going from 100 to 0

**Comment by InfusOnWoW** [comment]
I can't reproduce https://github.com/WeakAuras/WeakAuras2/pull/2000#issuecomment-590084776,

can you try running /dump WeakAuras.GetTriggerStateForTrigger("New 2", 1) ?

**Comment by mrbuds** [comment]
> /dump WeakAuras.GetTriggerStateForTrigger("New 2", 1)

https://i.imgur.com/oYiMlAm.jpg

It seems it has duplicated clones for same units with different unitid (player, partyX, raidX)

**Comment by mrbuds** [comment]
or easier to read https://i.imgur.com/vM8SNnE.png

**Comment by InfusOnWoW** [comment]
I hid those comments that I believe to be fixed now.

I can't reproduce any odditiy with the remaining time check.

I'll fix the classic compability and different directions tomorrow.

**Comment by mrbuds** [comment]
on deleting a text subregion
```
2x ...eakAurasOptions\SubRegionOptions\SubRegionCommon.lua:10: bad argument #1 to 'match' (string expected, got nil)
[string "=[C]"]: in function `match'
[string "@WeakAurasOptions\SubRegionOptions\SubRegionCommon.lua"]:10: in function <...eakAurasOptions\SubRegionOptions\SubRegionCommon.lua:6>
[string "@WeakAurasOptions\SubRegionOptions\SubRegionCommon.lua"]:41: in function `func'
[string "@WeakAuras\WeakAuras-@project-version@.lua"]:129: in function `ApplyToDataOrChildData'
[string "@WeakAurasOptions\SubRegionOptions\SubText.lua"]:49: in function <WeakAurasOptions\SubRegionOptions\SubText.lua:48>
[string "=[C]"]: ?
[string "@Ace3\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-78.lua"]:51: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:49>
[string "@Ace3\AceConfig-3.0-3\AceConfigDialog-3.0\AceConfigDialog-3.0-78.lua"]:843: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:664>
[string "=[C]"]: ?
[string "@Quartz\libs\AceGUI-3.0\AceGUI-3.0-40.lua"]:72: in function <Quartz\libs\AceGUI-3.0\AceGUI-3.0.lua:70>
[string "@Quartz\libs\AceGUI-3.0\AceGUI-3.0-40.lua"]:290: in function `Fire'
[string "@Ace3\AceGUI-3.0-40\widgets\AceGUIWidget-Icon.lua"]:26: in function <Ace3\AceGUI-3.0\widgets\AceGUIWidget-Icon.lua:25>
```

**Comment by mrbuds** [comment]
I can't reproduce issues with cast trigger's "remaining time" anymore

**Comment by InfusOnWoW** [comment]
The cast trigger issue being reversed will be fixed, but from my testing power triggers behave as before.

**Comment by mrbuds** [comment]
test on classic with master: https://i.imgur.com/YOCvrGe.jpg
with this PR: https://i.imgur.com/IEIoGf4.jpg

File: WeakAuras/GenericTrigger.lua [hunk 2]
```diff
@@ -533,6 +553,19 @@ local function RunTriggerFunc(allStates, data, id, triggernum, event, arg1, arg2
       else
         untriggerCheck = true;
       end
+    elseif (data.statesParameter == "unit") then
+      if arg1 then
+        allStates[arg1] = allStates[arg1] or {};
+        local state = allStates[arg1];
+        local ok, returnValue = xpcall(data.triggerFunc, errorHandler, state, event, arg1, arg2, ...);
+        if( (ok and returnValue) or optionsEvent) then
```
Comment (Stanzilla) [inline WeakAuras/GenericTrigger.lua]: style nit

File: WeakAuras/GenericTrigger.lua [hunk 3]
```diff
@@ -570,6 +603,19 @@ local function RunTriggerFunc(allStates, data, id, triggernum, event, arg1, arg2
             end
           end
         end
+      elseif data.statesParameter == "unit" then
+        if(data.untriggerFunc) then
```
Comment (Stanzilla) [inline WeakAuras/GenericTrigger.lua]: style nit

File: WeakAuras/GenericTrigger.lua [hunk 4]
```diff
@@ -570,6 +603,19 @@ local function RunTriggerFunc(allStates, data, id, triggernum, event, arg1, arg2
...
+          if arg1 then
+            local ok, returnValue =  xpcall(data.untriggerFunc, errorHandler, allStates[arg1], event, arg1, arg2, ...);
+            if(ok and returnValue) then
```
Comment (Stanzilla) [inline WeakAuras/GenericTrigger.lua]: style nit

File: WeakAuras/Types.lua [hunk 5]
```diff
@@ -2157,24 +2166,46 @@ WeakAuras.multiUnitId = {
...
+WeakAuras.multiUnitUnits = {
+  ["nameplate"] = {},
+  ["boss"] = {},
+  ["arena"] = {},
+  ["group"] = {},
+  ["party"] = {},
+  ["raid"] = {}
+}
+
+
```
Comment (Stanzilla) [inline WeakAuras/Types.lua]: one blank line too much

File: WeakAuras/Prototypes.lua [hunk 6]
```diff
@@ -1551,28 +1653,77 @@ WeakAuras.event_prototypes = {
...
+        name = "role",
+        display = L["Assigned Role"],
+        type = "select",
+        init = "UnitGroupRolesAssigned(unit)",
+        values = "role_types",
+        store = true,
+        conditionType = "select",
+        enable = function(trigger)
+           return not WeakAuras.IsClassic() and (trigger.unit == "group" or trigger.unit == "raid" or trigger.unit == "party")
+         end
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: There's a few "enable" function where the "end" is not aligned

**Pull request merged**